rpeach/weather-app
# Adding weather forecast functionality

Adding weather forecast functionality, replacing the demo app from create-react-app scaffolding.
This has been deployed to [github pages](https://rpeach.github.io/weather-app/).

The forecast is provided by [OpenWeatherMap](https://openweathermap.org/) with an api call limit of 60/second and 1,000,000/month (Currently using my hardcoded api key #11).

## Specification
The specification (including screen mockups and functionality) for this change was provided via an email document, but will not be uploaded here (Hopefully the reviewer has access to this).

## Features Added
* The app now has a searchbar to find the forecast at a particular location
* When entering data into the searchbar (and clicking search or pressing enter), the forecast info at that location is displayed.
* The time of the forecast can be adjusted with a slider
* The user can choose between temperature display in Celsius or Fahrenheit

## Testing
Some very basic automated tests were added in `src/components/App.test.tsx`, these have been run locally (this project doesn't yet use a CI server #8).

Some extra automated testing (particularly of the Weather component #7) would be useful in the future but was out of scope for this change.

## Manual testing
The following tests were run in IE11, Chrome (Stable, Latest), Firefox (Stable, Latest), Safari iOS (iPhone 8, Stable, Latest), Chrome Android (Samsung S8, Stable Latest) on Windows 10 (unless otherwise stated).
1. Launch website
2. **Check only Search bar and search button visible**
3. Enter location ("Cambridge, UK")
4. Click "Search" button
5. **Check that map displays correct location**
6. **Check that weather icon, temp and pressure display (Correct layout)**
7. Switch between Celsius and Fahrenheit
8. **Check that Temperature updates**
8. Drag slider to adjust forecast time
9. **Repeated above checks**

The above checks passed on all those browsers with a small display issue on IE11 (#2)

## Dependencies Added <sub>(Not including devDependencies)</sub>
|Dependency      |License        |Usage Description                                            |
|------------------|-------------|---------------------------------------------------|
| `@material-ui/...` | MIT            | Provides theming, components, styling         |
| `classnames`        | MIT           | A widely used tool for concatenating html element class names in react |
| `leaflet`                | BSD-2-Clause | Map - Tiles provided by [openstreetmap](https://www.openstreetmap.org/copyright) and appropriately attributed|
| `react-leaflet`      | MIT                 | React wrapper for leaflet                           |
| `Roboto font`     | Apache-2        | Font used by `@material-ui` |

**Note:** I have not checked the dependency tree for any unacceptable licenses

## Future Enhancements
* #1 The slider should display the day of the week rather than just a time offset
* #3 Store the current search location in the url
* #4 Save the temperature unit selected by the user
* #5 Show warning to the user on the entry of an invalid location
* Various code enhancements added to the issues list
